numpy/core/src/common/npy_longdouble.c

```diff
@@ -94,3 +95,77 @@ npy_longdouble_to_PyLong(npy_longdouble ldval)
     Py_DECREF(l_chunk_size);
     return v;
 }
+
+/* Helper function to get unicode(PyLong).encode('utf8') */
+static PyObject *
+_PyLong_Bytes(PyObject *long_obj) {
+    PyObject *bytes;
+#if defined(NPY_PY3K)
+    PyObject *unicode = PyObject_Str(long_obj);
+    if (unicode == NULL) {
+        return NULL;
+    }
+    bytes = PyUnicode_AsUTF8String(unicode);
+    Py_DECREF(unicode);
+#else
+    bytes = PyObject_Str(long_obj);
+#endif
+    return bytes;
+}
+
+
+/**
+ * TODO: currently a hack that converts the long through a string. This is
+ * correct, but slow.
+ *
+ * Another approach would be to do this numerically, in a similar way to
+ * PyLong_AsDouble.
+ * However, in order to respect rounding modes correctly, this needs to know
+ * the size of the mantissa, which is platform-dependent. To see why, consider
+ * converting
+ */
+NPY_VISIBILITY_HIDDEN npy_longdouble
+npy_longdouble_from_PyLong(PyObject *long_obj) {
+    npy_longdouble result;
+    char *end;
+    char *cstr;
+    PyObject *bytes;
+
+    bytes = _PyLong_Bytes(long_obj);
+    if (bytes == NULL) {
+        return -1;
+    }
+
+    cstr = PyBytes_AsString(bytes);
+    end = NULL;
+    errno = 0;
+    result = NumPyOS_ascii_strtold(cstr, &end);
+    Py_DECREF(bytes);
+    if (errno == ERANGE) {
+        /* strtold returns INFINITY of the correct sign. */
+        if (PyErr_Warn(PyExc_RuntimeWarning,
+                "overflow encountered in conversion from python long") < 0) {
+            return -1;
+        }
+    }
+    else if (errno) {
+        PyErr_Format(PyExc_RuntimeError,
+                     "Could not parse python long as longdouble: %s (%s)",
+                     cstr,
+                     strerror(errno));
+        return -1;
+    }
+
+    /* Extra characters at the end of the string, or nothing parsed */
+    if (end == cstr || *end) {
+        PyErr_Format(PyExc_RuntimeError,
+                     "Could not parse long as longdouble: %s",
+                     cstr);
+        return -1;
+    }
+
+    // Without this line, MSVC produces garbage (optimizes out result!?)
+    printf("Double form is %f\n", (double) result);
```

Comment:
I kid you not. Without this, `np.longdouble(0)` returns `-104892023.0`.

Reply:
In even more frightening news, replacing this line with:
```
// Without this line, MSVC produces garbage (optimizes out result!?)
printf("abc\n");
return result;
```

Causes `np.longdouble(x)` to always return `4.0`, which is `len('abc\n')`.

Reply:
could the compiler output from [godbot](https://godbolt.org/) help work out what is going on?

Reply:
This appears to be because the function prototype was not available before the function was called in `arraytypes.c.src`, which caused just as much chaos on my Mac until I added the appropriate `include` there, which I've now pushed in with my initial adjustments.

I suggest we might benefit from `warning: implicit declaration of function` into a compiler error--looks like this warning is effectively understating a break in C99 standard.